### ISSUE TITLE
MenuInflater tries to find onClick on wrong class with wrapped Context

### ISSUE THREAD

**Issue by grassick** [JakeWharton/ActionBarSherlock#533]
When using Theme.Sherlock.Light.DarkActionBar on Android 2.3.3, not with the other themes, the onClick of an action bar menu fails to bind (see error stack at bottom).

This is because the menu inflater is getting a wrapped context from ActionBarImpl:

``` java
if (targetThemeRes != 0) { //XXX && mContext.getThemeResId() != targetThemeRes) {
    mThemedContext = new ContextThemeWrapper(mContext, targetThemeRes);
} else {
    mThemedContext = mContext;
}
```

This then does not have the onClick method, since it is just a wrapper:

``` java
 public InflatedOnMenuItemClickListener(Context context, String methodName) {
    mContext = context;
    Class<?> c = context.getClass();
    try {
        mMethod = c.getMethod(methodName, PARAM_TYPES);
    } catch (Exception e) {
```

The InflatedOnMenuItemClickListener should check if it is a ContextWrapper and if so, get the base context, or the context passed in to it should already be the base context.

```
07-01 13:10:37.060: E/AndroidRuntime(3224): FATAL EXCEPTION: main
07-01 13:10:37.060: E/AndroidRuntime(3224): android.view.InflateException: Couldn't resolve menu item onClick handler onNewSource in class android.view.ContextThemeWrapper
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater$InflatedOnMenuItemClickListener.<init>(MenuInflater.java:204)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater$MenuState.setItem(MenuInflater.java:410)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater$MenuState.addItem(MenuInflater.java:445)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater.parseMenu(MenuInflater.java:175)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater.inflate(MenuInflater.java:97)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.github.androidimageprocessing.bacteria.ui.SourcesActivity.onCreateOptionsMenu(SourcesActivity.java:55)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at android.support.v4.app._ActionBarSherlockTrojanHorse.onCreatePanelMenu(_ActionBarSherlockTrojanHorse.java:45)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.ActionBarSherlock.callbackCreateOptionsMenu(ActionBarSherlock.java:556)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.internal.ActionBarSherlockCompat.preparePanel(ActionBarSherlockCompat.java:483)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.internal.ActionBarSherlockCompat.dispatchInvalidateOptionsMenu(ActionBarSherlockCompat.java:273)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.internal.ActionBarSherlockCompat$1.run(ActionBarSherlockCompat.java:988)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at android.os.Handler.handleCallback(Handler.java:587)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at android.os.Handler.dispatchMessage(Handler.java:92)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at android.os.Looper.loop(Looper.java:123)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at android.app.ActivityThread.main(ActivityThread.java:3687)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at java.lang.reflect.Method.invokeNative(Native Method)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at java.lang.reflect.Method.invoke(Method.java:507)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:842)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:600)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at dalvik.system.NativeStart.main(Native Method)
07-01 13:10:37.060: E/AndroidRuntime(3224): Caused by: java.lang.NoSuchMethodException: onNewSource
07-01 13:10:37.060: E/AndroidRuntime(3224):     at java.lang.ClassCache.findMethodByName(ClassCache.java:247)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at java.lang.Class.getMethod(Class.java:962)
07-01 13:10:37.060: E/AndroidRuntime(3224):     at com.actionbarsherlock.view.MenuInflater$InflatedOnMenuItemClickListener.<init>(MenuInflater.java:202)
07-01 13:10:37.060: E/AndroidRuntime(3224):     ... 19 more
```

**Comment by JakeWharton** [comment]
What happens when run on ICS+?

**Comment by grassick** [comment]
Emulator 4.0.3 gives exact same crash.

On Sun, Jul 1, 2012 at 1:44 PM, Jake Wharton <
reply@reply.github.com

**Comment by iNoles** [comment]
I am curious about Caused by: java.lang.NoSuchMethodException: onNewSource

**Comment by JakeWharton** [comment]
As I expected, this is also bug natively since it will also wrap with a `ContextThemeWrapper`. `onClick` attributes are not compile-time safe so I would recommend you specify them programmatically in your code.

**Comment by grassick** [comment]
I switched it to some uglier but functional manual adding of handlers to each item and it works. Would it be possible to add the single if statement to make it get the base context? It would prevent other code from mysteriously breaking when the theme is switched, as onclick isn't deprecated.

Thanks for the great library! - Clayton

**Comment by JakeWharton** [comment]
It would still throw an exception when running on Android 4.0 and above since we can't control that code. In my opinion, consistency across the native and custom action bar implementations is more important in this case.

**Comment by grassick** [comment]
Ah, but it works fine when implemented as native 4.0 code (I just tested
it). It's only when you use Sherlock classes that it crashes in 4.0.

On Sun, Jul 1, 2012 at 3:47 PM, Jake Wharton <
reply@reply.github.com

**Comment by JakeWharton** [comment]
Hmm... I'll look into it then.
